src/sentry/rules/base.py

```diff
@@ -108,6 +108,14 @@ def passes_activity(
     ) -> bool:
         raise NotImplementedError
 
+    def passes(self, event: GroupEvent, state: EventState) -> bool:
+        raise NotImplementedError
+
+    def after(
+        self, event: GroupEvent, state: EventState, notification_uuid: Optional[str] = None
+    ) -> Generator[CallbackFuture, None, None]:
+        raise NotImplementedError
```

Comment:
@asottile-sentry this is what ended up working for both the tests and mypy. let me know if you think this could be improved upon! i know we use this pattern a few other times for base classes in the codebase as well

Reply:
it would maybe be better to check your particular subclass:

```python
if not issubclass(whatever_cls, EventWhatever):
    raise TypeError(f'expected ...')
```

mypy understands this as a narrowing and will typecheck as `EventWhatever` afterward

rather than adding abc methods to something where a good percentage of the subclasses (intentionally) don't implement them

Reply:
I attempted this way in this [commit](https://github.com/getsentry/sentry/pull/58281/commits/021d19f8af268791829ae1106e6005323dd7c15e), but it had caused multiple tests to fail.

Reply:
Yeah I had a feeling we'd need to update them 😅  I'll look into them more tomorrow or next week, thanks so much!